src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -666,3 +667,109 @@ impl<T: Timestamp> CollectionState<T> {
         }
     }
 }
+
+/// A wrapper that holds on to backing persist shards/collections that the
+/// storage controller is aware of. The handles hold back the since frontier and
+/// we need to downgrade them when the read capabilities change.
+///
+/// Internally, this has an async task and the methods for registering a handle
+/// and downgrading sinces add commands to a queue that this task is working
+/// off. This makes the methods non-blocking and moves the work outside the main
+/// coordinator task, meaning the coordinator is spending less time waiting on
+/// persist calls.
+#[derive(Debug)]
+struct PersistWorker<T: Timestamp + Lattice + Codec64> {
+    tx: UnboundedSender<PersistWorkerCmd<T>>,
```

Comment:
if this was a mutex with some state instead of a channel (plus some sort of condvar), we could keep track of the highwater. this would also work for your `moving the periodic "advance frontier" calls to a separate
task and not use append() with empty updates for that` idea too.